### DISCUSSION TITLE
Coredns instances synchronization. One coredns answers the new POD IP and the other answers with the old POD IP.

### DISCUSSION THREAD

**Discussion by fcmercer** [coredns/coredns#6001] [Q&A] [answered]
Hello,
I have a normal starlingX AIO-DX installation with 2 coredns instances:

```
# kubectl get pod -n kube-system -o wide | grep "coredns"
coredns-594d9d85cf-qzdz9                          1/1     Running     0          5m5s    172.16.192.97    controller-0
coredns-594d9d85cf-z8htq                          1/1     Running     0          5m5s    172.16.166.180   controller-1
```

The scenario is:
A stateful POD acts as a server for a client POD.
The client POD sends some requests for the stateful POD but first, it sends a DNS lookup to resolve the IP.
If the stateful POD restarts, its IP will change.
During this change, the client POD is still sending the DNS lookup to send the requests. ( both coredns may send the old IP for a while due to the cache )
When the stateful POD is ready, one of the coredns can detects the new IP faster than the other.
Since the kube-dns is being used as a selector for this scenario, the client POD DNS lookup may be sent to coredns-594d9d85cf-qzdz9 or to coredns-594d9d85cf-z8htq
So the client POD receives different IP address of the stateful POD for a while.

```
i.e:
The stateful POD uses the IP: 10.1.1.1
After it restarts the new IP is: 10.1.1.2

The coredns-594d9d85cf-qzdz9 already detected this change,
but coredns-594d9d85cf-z8htq no.

Client POD receives:

dig stateful-pod
IP: 10.1.1.1 ( answered by coredns-594d9d85cf-z8htq )

dig stateful-pod
IP: 10.1.1.2  ( answered by coredns-594d9d85cf-qzdz9 )

dig stateful-pod
IP: 10.1.1.2  ( answered by coredns-594d9d85cf-qzdz9 )

dig stateful-pod
IP: 10.1.1.1  ( answered by coredns-594d9d85cf-z8htq )

After a while both coredns answers the new IP:

dig stateful-pod
IP: 10.1.1.2

dig stateful-pod
IP: 10.1.1.2

dig stateful-pod
IP: 10.1.1.2
```

The coredns version is: 1.8.0 and the config is:
```
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
            ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
            policy sequential
        }
        cache 30 {
            denial 9984 5
        }
        loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  ...
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
   ...
```

I am new on coredns.
I may change the config of cache or kubernetes plugins, but actually I was wondering how the coredns detects the new IP?
Does anyone have a clue where to look?

Trying to simulate this scenario, I could see this behavior just for some milliseconds.
![dns](https://user-images.githubusercontent.com/25907278/228869364-09618298-04cd-4ffb-8734-0cc7486dfaf7.jpg)

For me this behavior is normal, but Is there a way to avoid it? ( workaround )
I added the debug and log plugin to the coredns config, but I have no clue yet.

Best Regards and thanks,
Fabiano

**Comment by chrisohaver** [comment, marked as answer]
CoreDNS is informed of service object changes (e.g. service IP changes) via a Kubernetes API informer, and CoreDNS keeps these in memory (different than dns cache). Each CoreDNS instance is informed directly from the K8s API in this way, so it's possible that there could be some small amount of delay of updates between each instance.

However, cache is also independently maintained and not synchronized between instances. Records do not enter cache until a query for the record is received from a client.  So, it's possible that one instance has the item in cache, and another does not.  Or that each instance entered the item in cache at different times with different TTL remaining.  Or even that one cache instance still has the older record in it, and the other has the newer record.

**Reply by chrisohaver** [reply]
A workaround the cache discrepancy issue would be to disable cache for the kubernetes zone (cluster.local).
There are a few ways to do this...
* disable cache entirely for all zones
* disable just for cluster.local, by splitting the cluster.local domain and the reverse domains for service/pod ips into a separate server block that does not have cache plugin. 
* disable just for cluster.local, by updating to a coredns version that supports the  `disable success|denial [ZONES...]` option in cache to disable success and denial cache for cluster.local. (and service/pod ip reverse zones).

**Reply by fcmercer** [reply]
Hello @chrisohaver  thanks for your answer.
Yesterday I tested the coredns 1.10.1 with:
```
cache {
     ...
     disable success xxxx.cluster.local
     disable denial xxxx.cluster.local
}
```
Then I created a statefulset POD running in xxxx.cluster.local and another one running on yyyy.cluster.local
It worked fine, the cache was still working for yyyy.cluster.local but not for xxxx.cluster.local 

But it seems we can't upgrade to coredns 1.10.1 yet.
So I will test your second option:

`disable just for cluster.local, by splitting the cluster.local domain and the reverse domains for service/pod ips into a separate server block that does not have cache plugin.`

I am not sure if I understood the case: Should I just configure the coredns config like this?
```
xxxx.cluster.local {
     # no cache config
}
yyyy.cluster.local {
     cache {
     }
}
```
But if the idea above is correct, it means I will need to specify the cache for all domains individually?
Is there a way to configure the cache config for all other domains and just overwrite for cluster.local? i.e:

```
# the cache config for all domains
cache {
   ....
}
xxxx.cluster.local {
     # disable the cache config for this domain.....how?
}
```
I will try to do these configs now

**Comment by chrisohaver** [comment]
> But it seems we can't upgrade to coredns 1.10.1 yet.

the option was added in 1.9.4

**Comment by chrisohaver** [comment]
The cache exclude would look like ...

```
cache 30 {
     disable success cluster.local 96.10.in-addr.arpa 172.in-addr.arpa
     disable denial cluster.local 96.10.in-addr.arpa 172.in-addr.arpa
}
```

Note the entire k8s cluster domain is excluded from cache... not `xxxx` .
Also note the reverse domains included here are _just examples_.  Your service and pod ip ranges may (will) be different.


The two server block  example would look something like ...

```
. {
        errors
        cache 30 {
            denial 9984 5
        }
        prometheus :9153
        forward . /etc/resolv.conf {
            policy sequential
        }
        loop
        reload
        loadbalance
}

cluster.local 96.10.in-addr.arpa 172.in-addr.arpa {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes {
            pods insecure
            ttl 30
        }
        loadbalance
}
```

**Reply by fcmercer** [reply]
I have just tested this dns config: ( I added the service name in the domain )

```
apiVersion: v1
data:
  Corefile: |
    centos-svc.default.svc.cluster.local:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
            ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
            policy sequential
        }
        loop
        reload
        loadbalance
    }
    .:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
            ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
            policy sequential
        }
        cache 30 {
            denial 9984 5
        }
        loop
        reload
        loadbalance
    }
```

I created 2 services:
centos-svc.default.svc.cluster.local
and
centos-svc2.default.svc.cluster.local

I deleted the statefulset pod on centos-svc2 and coredns was still sending the its old IP from cache
but when I deleted the the statefulset pod on centos-svc the coredns was not sending the OLD IP anymore.

I think it will solve my issue.
I haven't added the 96.10.in-addr.arpa 172.in-addr.arpa to the config. I will try to understand it now.

**Comment by fcmercer** [comment]
Thanks @chrisohaver you helped me a lot.
